File: 백준/Silver/1992. 쿼드트리/쿼드트리.cc
#include<bits/stdc++.h>
using namespace std;
const int dy[] = {-1,0,1,0};
const int dx[] = {0,1,0,-1};
int a[68][68];
int n;
void DFS(int y,int x,int size)
{
	int val=a[y][x];
	bool same = true;
	for(int i=y;i<y+size;i++)
	{
		for(int j=x;j<x+size;j++)
		{
			if(a[i][j]!=val)
			{
				same = false;
				break;
			}
		}
		if(!same) break;
	}
	
	if(same)
	{
		cout<<val;
		return;
	}
	
	int half = size/2;
	cout<<'(';
	DFS(y,x,half);
	DFS(y,x+half,half);
	DFS(y+half,x,half);
	DFS(y+half,x+half,half);
	cout<<')';
}
int main()
{
	scanf("%d",&n);
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<n;j++)
		{
			scanf("%1d",&a[i][j]);
		}
	}
	DFS(0,0,n);
}

File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
def dfs(y,x,size):
    val=a[y][x]
    for i in range(y,y+size):
        for j in range(x,x+size):
            if val!=a[i][j]:
                half=size//2
                print('(',end='')
                dfs(y,x,half)
                dfs(y,x+half,half)
                dfs(y+half,x,half)
                dfs(y+half,x+half,half)
                print(')',end='')
                return
    print(val,end='')

n=int(input().strip())
a=[[0]*n for _ in range(n)]
for i in range(n):
    line=input().strip()
    for j in range(n):
        a[i][j]=int(line[j])

dfs(0,0,n) #y,x,size